[Merged by Bors] - feat(CategoryTheory/Sites): global sections functor on sheaves

peabrainiac opened the pull request:
This PR defines `Sheaf.Γ : Sheaf J A ⥤ A` as the functor taking each sheaf to the limit of its underlying presheaf whenever possible, i.e. whenever `A` has sufficiently large limits, and shows that this agrees with several other common definitions when they are defined, like evaluation of sheaves on a terminal object. See [Zulip](https://leanprover.zulipchat.com/#narrow/channel/116395-maths/topic/global.20sections.20functor) for the discussion that lead to this approach.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->
My goal with this PR is to work towards the cohesive structure of sheaves on cohesive sites; I have already defined local sites [here](https://github.com/peabrainiac/lean-orbifolds/blob/main/Orbifolds/ForMathlib/LocalSite.lean) as a stepping stone and will PR them too shortly.


[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

peabrainiac commented:
The main thing I am unsure about here is where to place the prerequisite lemmas that do not involve `Sheaf.Γ` already; for `limNatIsoSectionsFunctor ` and `sheafSectionsNatIsoEvaluation` I think I have found good homes, and `terminalObjIso` could be placed in `Limits.FunctorCategory.Shapes.Products` too with only a minor increase in imports, but for `sectionsFunctorNatIsoCoyoneda` I have not been able to find a good home other than the global sections file itself. I suppose we could just leave it there, but maybe someone else has a better idea.

peabrainiac commented on Mathlib/CategoryTheory/Sites/Sheaf.lean:
```diff
@@ -345,6 +345,11 @@ def sheafToPresheaf : Sheaf J A ⥤ Cᵒᵖ ⥤ A where
 abbrev sheafSections : Cᵒᵖ ⥤ Sheaf J A ⥤ A := (sheafToPresheaf J A).flip
 
+/-- The sheaf sections functor on `X` is given by evaluation of presheaves on `X`. -/
+def sheafSectionsNatIsoEvaluation {X : C} :
+    (sheafSections J A).obj (op X) ≅ sheafToPresheaf J A ⋙ (evaluation _ _).obj (op X) :=
+  NatIso.ofComponents fun _ ↦ eqToIso rfl
```
Oh, also this seems like just a general lemma about `Functor.flip` - do we want that in its general form? Or do we maybe even have it already, and I just couldn't find it?

joelriou replied:
```suggestion
@[simps!]
def sheafSectionsNatIsoEvaluation {X : C} :
    (sheafSections J A).obj (op X) ≅ sheafToPresheaf J A ⋙ (evaluation _ _).obj (op X) :=
  NatIso.ofComponents (fun _ ↦ Iso.refl _)
```

joelriou commented on Mathlib/CategoryTheory/Sites/GlobalSections.lean:
```diff
@@ -0,0 +1,95 @@
+
+/-- For sheaves of types, the global sections functor is isomorphic to the covariant hom
+functor of the terminal sheaf. -/
+noncomputable def Sheaf.ΓNatIsoCoyonedaObj :
+    Γ J (Type max u w) ≅ coyoneda.obj (op (⊤_ _)) := by
```
This should be first stated as an equivalence of types for each sheaf (with more general universes, the LHS and RHS will be in different universes), followed by a naturality lemma, and then it can be stated as a natural isomorphism when the universes match.

peabrainiac replied:
> This should be first stated as an equivalence of types for each sheaf (with more general universes, the LHS and RHS will be in different universes), followed by a naturality lemma, and then it can be stated as a natural isomorphism when the universes match.

I'm unsure about what exactly you're suggesting - should all the natural isomorphisms here be stated in terms of isomorphisms/equivalences first, or is there some special reason to do it specifically with this one? I would have thought that isomorphisms involving `Γ` are particularly bad candidates for a more step-by-step API like this, since `Γ` is now defined simply as a right-adjoint and not really constructed anyways.

joelriou replied:
I am referring only to natural "isomorphisms" between functors to types, where the target types may live in different universes, which happens here when universes are left completely general.

joelriou commented on Mathlib/CategoryTheory/Sites/GlobalSections.lean:
```diff
@@ -0,0 +1,95 @@
+variable {C : Type u} [Category.{v} C] (J : GrothendieckTopology C)
+  (A : Type u₂) [Category.{v₂} A]
+
+/-- The global sections functor `Γ`, taking sheaves valued in `A` to its global sections in `A`.
+It is given here as a limit instead of evaluation on a terminal object of `C`; this has the
+advantage of not requiring `C` to have a terminal object, but the disadvantage of requiring `A`
+to have limits of shape `Cᵒᵖ`. -/
+noncomputable def Sheaf.Γ [HasLimitsOfShape Cᵒᵖ A] : Sheaf J A ⥤ A :=
+  sheafToPresheaf J A ⋙ lim
```
Unfortunately, I doubt that the interesting cases in algebraic geometry where there is a global section functor but no terminal object are covered by these assumptions. Sites in algebraic geometry tend to be large, and we may have a small family of objects which cover the terminal object even though we may not have limits indexed by `Cᵒᵖ`. It means that we would probably have to refactor this definition in the future. This is part of the reason why I had suggested another design where we define the global sections by taking a right adjoint of the constant sheaf functor. I have just drafted the following (which also takes advantage of the already known results about the constant sheaf functor, see the file `Sites.ConstantSheaf`), and it does not seem to be difficult to obtain the same results which you obtained with the other definition:

```lean
variable [HasWeakSheafify J A]

abbrev HasGlobalSectionsFunctor := (constantSheaf J A).IsLeftAdjoint

instance hasGlobalSectionsFunctor_of_hasTerminal [HasTerminal C] :
    HasGlobalSectionsFunctor J A :=
  ⟨_, ⟨constantSheafAdj J A terminalIsTerminal⟩⟩

section

variable [HasGlobalSectionsFunctor J A]

noncomputable def Sheaf.Γ : Sheaf J A ⥤ A :=
  (constantSheaf J A).rightAdjoint

/-- The constant sheaf functor is left-adjoint to the global sections functor when both exist. -/
noncomputable def constantSheafΓAdj :
    constantSheaf J A ⊣ Γ J A :=
  Adjunction.ofIsLeftAdjoint (constantSheaf J A)

instance : (constantSheaf J A).IsLeftAdjoint :=
  ⟨Γ J A, ⟨constantSheafΓAdj J A⟩⟩

instance : (Γ J A).IsRightAdjoint :=
  ⟨constantSheaf J A, ⟨constantSheafΓAdj J A⟩⟩

variable {J A}

noncomputable def Sheaf.ΓHomEquiv {X : A} {F : Sheaf J A} :
    ((Functor.const _).obj X ⟶ F.val) ≃ (X ⟶ (Γ J A).obj F) :=
  ((sheafificationAdjunction J A).homEquiv _ _).symm.trans
    ((constantSheafΓAdj J A).homEquiv _ _)

@[simps pt]
noncomputable def Sheaf.coneΓ (F : Sheaf J A) : Cone F.val where
  pt := (Γ J A).obj F
  π := ΓHomEquiv.symm (𝟙 _)

def Sheaf.isLimitConeΓ (F : Sheaf J A) : IsLimit F.coneΓ := by
  -- should be easy from `ΓHomEquiv` and a few naturality lemmas
  sorry

noncomputable def Sheaf.Γres (F : Sheaf J A) (U : Cᵒᵖ) :
  (Γ J A).obj F ⟶ F.val.obj U := F.coneΓ.π.app U

@[reassoc (attr := simp)]
lemma Sheaf.Γres_map (F : Sheaf J A) {V U : Cᵒᵖ} (f : U ⟶ V) :
    F.Γres U ≫ F.val.map f = F.Γres V :=
  F.coneΓ.w f

@[simp]
lemma Sheaf.coneΓ_π_app (F : Sheaf J A) (U : Cᵒᵖ) : F.coneΓ.π.app U = F.Γres U := rfl

noncomputable def Sheaf.ΓNatIsoSheafSections {T : C}
    (hT : IsTerminal T) : Γ J A ≅ (sheafSections J A).obj (op T) :=
  Adjunction.rightAdjointUniq (constantSheafΓAdj J A) (constantSheafAdj J A hT)

end
```

peabrainiac commented:
Thanks for your review and explanation of why this extra generality is actually important / useful. I've rewritten this in terms of `HasGlobalSectionsFunctor` now like you suggested, and also added `Sheaf.ΓNatIsoLim` to show that this is equivalent to the previous definition when `A` has the right limits. I didn't go through the route via `Sheaf.ΓHomEquiv` that you outlined yet, but can also try implementing that tomorrow (I just wanted to get this refactor pushed first).

One slightly annoying thing I noticed is that with this new definition, `Sheaf.Γ` now requires `HasWeakSheafify`, which [`instHasSheafifyOfHasFiniteLimits`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Sites/LeftExact.html#CategoryTheory.instHasSheafifyOfHasFiniteLimits) only seems to provide for `Type max u v` but not higher universe levels like `Type max u v w`; at least I couldn't get it to work with the latter, and so had to remove `w` from the last few lemmas in the file. In the cases I am most interested in `u` and `v` are both 0, so this is quite a loss of generality. This does seem more like an issue with the instance than a mathematical issue to me though, so hopefully it can be resolved with another PR.

joelriou commented on Mathlib/CategoryTheory/Sites/GlobalSections.lean:
```diff
@@ -0,0 +1,116 @@
+    Γ J A ≅ sheafToPresheaf J A ⋙ lim :=
+  (constantSheafΓAdj J A).rightAdjointUniq (constLimAdj.comp (sheafificationAdjunction J A))
+
+-- this is currently needed to obtain the instance `HasSheafify J (Type max u v)`.
+attribute [local instance] CategoryTheory.Types.instConcreteCategory
+attribute [local instance] CategoryTheory.Types.instFunLike
+
+/-- For functors `F : C ⥤ Type _`, `F.sections` is naturally isomorphic to the type `⊤_ _ ⟶ F`
+of natural transformations from the terminal functor to `F`. -/
+noncomputable def sectionsFunctorNatIsoCoyoneda :
```
This is not directly related to sections of sheaves as this does not involve a Grothendieck topology. This should be moved in another file, and as I suggested, it should be first stated as a bijection (+ a naturality statement):
```lean
@[simps apply_app]
def Functor.sectionsEquivHom (F : C ⥤ Type w) (X : Type w) [Unique X] :
    F.sections ≃ ((Functor.const _).obj X ⟶ F) where
  toFun s :=
    { app j x := s.1 j
      naturality _ _ _ := by ext x; simp }
  invFun τ := ⟨fun j ↦ τ.app _ (default : X), fun φ ↦ (congr_fun (τ.naturality φ) _).symm⟩
  left_inv s := rfl
  right_inv τ := by
    ext j (x : X)
    obtain rfl := Subsingleton.elim x default
    rfl
```
The sheaf version would be:
```lean
variable {J} in
noncomputable def Sheaf.ΓObjEquiv [HasWeakSheafify J (Type w)]
    [HasGlobalSectionsFunctor J (Type w)] (F : Sheaf J (Type w)) :
      (Γ J (Type w)).obj F ≃ F.1.sections :=
  (Equiv.trans (by exact (Equiv.funUnique PUnit _).symm)
    Sheaf.ΓHomEquiv.symm).trans (Functor.sectionsEquivHom F.1 PUnit).symm
```

peabrainiac commented:
That should be the last of the requested changes implemented now, I think.

joelriou commented on Mathlib/CategoryTheory/Sites/GlobalSections.lean:
```diff
@@ -0,0 +1,214 @@
+is isomorphic to several other constructions when they exist, most notably evaluation of sheaves
+on a terminal object and `Functor.sectionsFunctor`.
+
+## Main definitions / results
```
Could you update this section as the names do not necessarily match the definitions?

peabrainiac replied:
I think `ΓNatIsoCoyonedaObj` was the only mismatch; let me know if there's more. I've also added `isLimitConeΓ` and `Γres` to that section now.

joelriou commented on Mathlib/CategoryTheory/Sites/GlobalSections.lean:
```diff
@@ -0,0 +1,216 @@
+
+/-!
+# Global sections of sheaves
+In this file we define a global sections functor `Sheaf.Γ : Sheaf J A ⥤ A` and show that it
```
```suggestion

In this file we define a global sections functor `Sheaf.Γ : Sheaf J A ⥤ A` and show that it
```

joelriou commented on Mathlib/CategoryTheory/Yoneda.lean:
```diff
@@ -797,6 +797,44 @@ lemma yonedaMap_app_apply {Y : C} {X : Cᵒᵖ} (f : X.unop ⟶ Y) :
+  rfl
+
+lemma Functor.sectionsEquivHom_naturality_symm {F G : C ⥤ Type u₂} (f : F ⟶ G) (X : Type u₂)
+    [Unique X] (τ : (const C).obj X ⟶ F) : (G.sectionsEquivHom X).symm (τ ≫ f) =
+    (sectionsFunctor C).map f ((F.sectionsEquivHom X).symm τ) := by
```
```suggestion
    [Unique X] (τ : (const C).obj X ⟶ F) :
    (G.sectionsEquivHom X).symm (τ ≫ f) =
      (sectionsFunctor C).map f ((F.sectionsEquivHom X).symm τ) := by
```

joelriou commented on Mathlib/CategoryTheory/Yoneda.lean:
```diff
@@ -797,6 +797,44 @@ lemma yonedaMap_app_apply {Y : C} {X : Cᵒᵖ} (f : X.unop ⟶ Y) :
+embedding of a terminal functor, specifically a constant functor on a given singleton type `X`. -/
+@[simps!]
+noncomputable def sectionsFunctorNatIsoCoyoneda (X : Type max u₁ u₂) [Unique X] :
+    Functor.sectionsFunctor.{v₁,max u₁ u₂} C ≅ coyoneda.obj (op ((Functor.const C).obj X)) :=
```
```suggestion
    Functor.sectionsFunctor.{v₁, max u₁ u₂} C ≅ coyoneda.obj (op ((Functor.const C).obj X)) :=
```

joelriou commented on Mathlib/CategoryTheory/Sites/GlobalSections.lean:
```diff
@@ -0,0 +1,216 @@
+
+/-- Global sections of sheaves are naturally isomorphic to the limits of the underlying presheaves.
+Note that while `HasLimitsOfShape Cᵒᵖ A` is needed here to talk about `lim` as a functor, global
+sections are always limits, it just has to be stated a little bit more carefully. -/
```
```suggestion
sections are always limits, it just has to be stated a little bit more carefully,
see `Sheaf.isLimitConeΓ`. -/
```

joelriou commented on Mathlib/CategoryTheory/Sites/GlobalSections.lean:
```diff
@@ -0,0 +1,216 @@
+@[simp]
+lemma Sheaf.coneΓ_π_app [HasGlobalSectionsFunctor J A] (F : Sheaf J A) (U : Cᵒᵖ) :
+    F.coneΓ.π.app U = F.Γres U := rfl
+
```
I think this PR is very close to being merged. Could you just add a lemma `Sheaf.Γres_naturality` expressing the naturality of `Γres` in `F`, and then define a natural transformation `Sheaf.natTransΓres : Γ J A ⟶ (sheafSections J A).obj U`?

peabrainiac replied:
I've added it as `natTransΓres` now, but I'm curious - what does the naming convention have to say about whether it should be `natTransΓres` or `ΓresNatTrans`? The latter is what I would have expected it to be. Which reminds me, I'm also a bit confused about why `Γres` is not called `ΓRes` when definitions are usually in camelCase and `Res` is (arguably) the second word in that variable name.

peabrainiac replied:
Changed it now.

peabrainiac commented:
> One slightly annoying thing I noticed is that with this new definition, `Sheaf.Γ` now requires `HasWeakSheafify`, which [`instHasSheafifyOfHasFiniteLimits`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Sites/LeftExact.html#CategoryTheory.instHasSheafifyOfHasFiniteLimits) only seems to provide for `Type max u v` but not higher universe levels like `Type max u v w`; at least I couldn't get it to work with the latter, and so had to remove `w` from the last few lemmas in the file. In the cases I am most interested in `u` and `v` are both 0, so this is quite a loss of generality. This does seem more like an issue with the instance than a mathematical issue to me though, so hopefully it can be resolved with another PR.

While applying the latest batch of suggestions I have also taken the opportunity to add a TODO note about this now.

joelriou commented on Mathlib/CategoryTheory/Sites/GlobalSections.lean:
```diff
@@ -0,0 +1,240 @@
+    [HasGlobalSectionsFunctor J (Type w)] {F G : Sheaf J (Type w)} (f : F ⟶ G)
+    (x : F.val.sections) : (ΓObjEquivSections J G).symm ((Functor.sectionsFunctor _).map f.val x) =
+      (Γ J _).map f ((ΓObjEquivSections J F).symm x) :=
+  (congr_fun (ΓHomEquiv_naturality_right (F.val.sectionsEquivHom _ x) f) _)
```
```suggestion
  congr_fun (ΓHomEquiv_naturality_right (F.val.sectionsEquivHom _ x) f) _
```

joelriou commented on Mathlib/CategoryTheory/Sites/GlobalSections.lean:
```diff
@@ -0,0 +1,240 @@
+
+/-- For sheaves of types, the global sections functor is isomorphic to the covariant hom
+functor of the terminal sheaf. -/
+noncomputable def Sheaf.ΓNatIsoCoyoneda (X : Type max u v) [Unique X] :
+    Γ J (Type max u v) ≅ coyoneda.obj (op ((constantSheaf J _).obj X)) := by
+  exact NatIso.ofComponents (fun F ↦ (F.ΓObjEquivHom J X).toIso) fun f ↦ by
+    ext x
+    exact ΓObjEquivHom_naturality J X f x
```
```suggestion
noncomputable def Sheaf.ΓNatIsoCoyoneda (X : Type max u v) [Unique X] :
    Γ J (Type max u v) ≅ coyoneda.obj (op ((constantSheaf J _).obj X)) :=
  NatIso.ofComponents (fun F ↦ (F.ΓObjEquivHom J X).toIso) fun f ↦ by
    ext x
    exact ΓObjEquivHom_naturality J X f x
```
